sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,626 @@
+        child.w[ax] = parent.w[ax] / 2.0
+        child.le[ax] = parent.le[ax] + offset[ax] * parent.w[ax] / 2.0
+        child.c[ax] = child.le[ax] + child.w[ax] / 2.0
+        child.cum_com[ax] = 0.
+        child.cur_pos[ax] = -1.
+    child.tree.num_cells += 1
+    return child
+
+cdef Node* select_child(Node *node, float[3] pos, long index) nogil:
+    # Find which sub-node a position should go into
+    # And return the appropriate node
+    cdef int[3] offset
+    cdef int ax
+    cdef Node* child
+    # In case we don't have 3D data, set it to zero
+    for ax in range(3):
+        offset[ax] = 0
+    for ax in range(node.tree.dimension):
+        offset[ax] = (pos[ax] - (node.le[ax] + node.w[ax] / 2.0)) > 0.
+        # If the points are arbitrarily close (within machine epsilon)
+        # on a given axis, then flip the offset on this axis.
+        # Only do this when we are considering different points 
+        # (as evidenced by the index)
+        if (pos[ax] == node.cur_pos[ax]) & (index != node.point_index):
+            offset[ax] = 1 - offset[ax]
+            if DEBUGFLAG:
+                printf("[t-SNE] flipped node.cur_pos %i [%f, %f] pos %i [%f, %f]\n", node.point_index, node.cur_pos[0], node.cur_pos[1], index, pos[0], pos[1])
+    child = node.children[offset[0]][offset[1]][offset[2]]
+    if DEBUGFLAG:
+        printf("[t-SNE] Offset [%i, %i] with LE [%f, %f]\n", offset[0], offset[1], child.le[0], child.le[1])
+    return child
+
+cdef void subdivide(Node* node) nogil:
+    # This instantiates 4 or 8 nodes for the current node
+    cdef int i = 0
+    cdef int j = 0
+    cdef int k = 0
+    cdef int[3] offset
+    node.is_leaf = False
+    for ax in range(3):
+        offset[ax] = 0
+    if node.tree.dimension > 2:
+        krange = 2
+    else:
+        krange = 1
+    for i in range(2):
+        offset[0] = i
+        for j in range(2):
+            offset[1] = j
+            for k in range(krange):
+                offset[2] = k
+                node.children[i][j][k] = create_child(node, offset)
+
+
+cdef int insert(Node *root, float pos[3], long point_index, long depth) nogil:
+    # Introduce a new point into the tree
+    # by recursively inserting it and subdividng as necessary
+    cdef Node *child
+    cdef long i
+    cdef int ax
+    cdef int dimension = root.tree.dimension
+    # Increment the total number points including this
+    # node and below it
+    root.cum_size += 1
+    # Evaluate the new center of mass, weighting the previous
+    # center of mass against the new point data
+    cdef double frac_seen = <double>(root.cum_size - 1) / (<double> root.cum_size)
+    cdef double frac_new  = 1.0 / <double> root.cum_size
+    for ax in range(dimension):
+        root.cum_com[ax] *= frac_seen
+        if (pos[ax] > (root.le[ax] + root.w[ax])):
+            printf("[t-SNE] Error: point (%1.9e) is above right edge of node (%1.9e)\n", pos[ax], root.le[ax] + root.w[ax])
```

Comment:
pep8 gives me 167 warnings for this file, I think this could be reduced significantly. Although we can ignore something like `_barnes_hut_tsne.pyx:458:14: E225 missing whitespace around operator` these are important:

* `_barnes_hut_tsne.pyx:192:80: E501 line too long (82 > 79 characters)`
* `_barnes_hut_tsne.pyx:407:47: E231 missing whitespace after ','`
* `_barnes_hut_tsne.pyx:442:1: E303 too many blank lines (3)`
* `_barnes_hut_tsne.pyx:481:46: W291 trailing whitespace`
* ...